api/src/verify_token.rs

```diff
@@ -0,0 +1,222 @@
+                return Ok(PossibleToken::Verified);
+            }
+
+            match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+                Ok(_) => return Ok(PossibleToken::Unverified),
+                Err(e) => {
+                    return Err(VerifyTokenError::Serde { source: e });
+                },
+            }
+        },
+        Err(_) => Ok(PossibleToken::Unverified),
+    }
+}
+
+pub async fn create(
+    db: &impl Database,
+    id: &str,
+    expires_at: DateTime<Utc>,
+) -> Result<UnverifiedToken>
+{
+    if let Ok(bytes) = db.get(&id).await {
+        if serde_json::from_slice::<VerifiedToken>(&bytes).is_ok() {
+            return Err(VerifyTokenError::Done);
+        }
+
+        match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+            Ok(_) => {},
+            Err(e) => {
+                return Err(VerifyTokenError::Serde { source: e });
+            },
+        };
+
+        // otherwise overwrite the unverified token, which corresponds
+        // to the case of "resend the confirmation email", etc.
+    }
+
+    let secret = Mu::new();
+    let encoded_secret = base64::encode(secret.0);
+
+    let t = UnverifiedToken {
+        schema_version: 0,
+        secret: encoded_secret,
+        expires_at: expires_at,
+    };
+
+    let bytes = match serde_json::to_vec(&t) {
+        Ok(b) => b,
+        Err(e) => return Err(VerifyTokenError::Serde { source: e }),
+    };
+
+    match db.put(id, &bytes).await {
+        Ok(_) => Ok(t),
+        Err(_) => Err(VerifyTokenError::Unknown),
+    }
+}
+
+pub async fn verify(
+    db: &impl Database,
+    id: &str,
+    secret: &str,
+    method: VerifyMethod,
+) -> Result<VerifiedToken>
+{
+    if let Ok(bytes) = db.get(id).await {
+        if serde_json::from_slice::<VerifiedToken>(&bytes).is_ok() {
+            return Err(VerifyTokenError::Done);
+        }
+
+        match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+            Ok(u) => match u.schema_version {
+                0 => {
+                    if Utc::now() > u.expires_at {
+                        return Err(VerifyTokenError::Expired);
+                    }
+
+                    if u.secret != secret {
+                        return Err(VerifyTokenError::Secret);
+                    }
+
+                    let v = VerifiedToken { schema_version: 0, method: method };
+
+                    let bytes = match serde_json::to_vec(&v) {
+                        Ok(b) => b,
+                        Err(e) => {
+                            return Err(VerifyTokenError::Serde { source: e });
+                        },
+                    };
+
+                    return match db.put(id, &bytes).await {
+                        Ok(_) => Ok(v),
+                        Err(_) => Err(VerifyTokenError::Unknown),
+                    };
+                },
+                _ => {
+                    return Err(VerifyTokenError::Schema);
```

Comment:
the way that it stands, if the schema changed, the data wouldn't even parse and we wouldn't be able to get to a point where we can read the schema version in the first place. we might have to parse it with a "meta model" that just looks at the schema version and then parse it for real with the full model

Reply:
It would parse fine, if i'm understanding you correctly. But we only support schema version 0 right now so anything else would be an error.

Reply:
if we updated the code and the model changed and serde didn't find a required key in the old version of the data, wouldn't we get a serde error?

Reply:
yes except we might change other things within the fields themselves, like the time format, or the length of the secret, etc.